Extending fluid concept to fit CF & Standard formulations

vlipovac opened the pull request:
Closes #1240
Closes #1247 

## Proposed changes

Reworking the concept of `model.fluid` to be a general fluid with phases and components, instead of just a data collection of constants.

Breaking change with a multiple minor changes throughout the packages. The essential changes are summarized in a self-review.

`porepy.compositional.ChemicalSpecies` was merged into `FluidConstants`, including data class functionality. This led to some changes for `SolidConstants` as well for inheritance reasons.

## Types of changes

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [X] Testing (contribution related to testing of existing or new functionality).
- [x] Documentation (contribution related to adding, improving, or fixing documentation).
- [X] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

- [X] The documentation is up-to-date.
- [X] Static typing is included in the update.
- [X] This PR does not duplicate existing functionality.
- [X] The update is covered by the test suite (including tests added in the PR).
- [X] `pytest` was run with the `--run-skipped` flag.

vlipovac commented on src/porepy/models/solution_strategy.py:
```diff
@@ -203,6 +214,12 @@ def prepare_simulation(self) -> None:
+        # NOTE this is critical in the case where properties depend on some fractions.
+        # The callables for those are dynamically created during create_variables, as opposed
+        # to e.g., pressure or temperature.
+        self.assign_thermodynamic_properties_to_phases()
```
This step was added to the set-up.
It is critical that the Fluid is created in `set_materials`, before the variables, since the dynamic number of variables depends on the fluid definition.
The step `assign_thermodynamic_properties_to_phases` requires the variables to be already created, since fluid properties are in general secondary terms which depend on some variable (keyword SurrogateFactory)

vlipovac commented on src/porepy/models/fluid_property_library.py:
```diff
@@ -0,0 +1,375 @@
+""""Module containing some constant heuristic fluid property implemenations.
```
This file takes fluid-related heuristics out of the `constitutive_laws.py` file.
Some methods are reworked to fit the general fluid concept, where properties of phases are dynamically created based on the number of modelled phases.

vlipovac commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1095,25 +1129,37 @@ def create_mixture(self) -> None:
-        self.fluid_mixture = FluidMixture(components, phases)
+        self.fluid = Fluid(components, phases)
 
     def get_components(self) -> list[Component]:
```
This method has now a default-behavior, creating a one-component fluid based on the fluid constants passed by the user. Important to mention is that the simulation set-up does not change for the standard 1-phase, 1-component case.
Users still use and pass a FluidConstants instance in params.

keileg replied:
Can be resolved for my sake.

vlipovac commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1128,17 +1174,14 @@ def get_phase_configuration(
-        raise CompositionalModellingError(
-            "Call to mixin method. Configure phases by overriding this method."
-        )
+        return [(AbstractEoS(components), PhysicalState.liquid, "liquid")]
```
The default behavior of get_phase_configuration was also changed to model a 1-phase fluid with a liquid-like phase.
The AbstractEoS was modified for this purpose, not being abstract anymore but playing the role of the "mathematician's" EoS, i.e. a dummy doing nothing for models using heuristic fluid properties (constitutive_laws.py)

keileg replied:
Can be resolved for my sake

vlipovac commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1171,42 +1211,30 @@ def assign_thermodynamic_properties_to_mixture(self) -> None:
-        self.fluid_mixture.specific_volume = self.fluid_specific_volume
-        self.fluid_mixture.specific_enthalpy = self.fluid_specific_enthalpy
-
     def dependencies_of_phase_properties(
```
This is a bit tricky.

I hade to change the default behavior of this, because of the mechanics-only models, who inherit from SolutionStrategy.
Problems include the missing of the classical variable mixins like pressure and temperature.

The default return value is now an empty list, leading to fluid properties which raise an error if they were not overridden by some constitutive law.
It makes sense, but may be hard to digest for people digging deeper into the code.

vlipovac replied:
Improved documentation, here and at the default implementation of fluid properties in this class.

keileg replied:
Can be resolved from my side

vlipovac commented on src/porepy/compositional/base.py:
```diff
@@ -957,3 +924,137 @@ def reference_component(self) -> Component:
         :meth:`reference_component_index`."""
         return self._components[self.reference_component_index]
+
+    def density(self, domains: pp.SubdomainsOrBoundaries) -> pp.ad.Operator:
```
The class `Fluid` (formerly FluidMixture) is now the new interface for other model mixin classes to access fluid properties.
It implements the fluid properties required by the PDE classes, while the framework handles the assignment of heuristic functions to the (single) phase properties under the hood.

vlipovac commented on tests/models/test_fluid_mass_balance.py:
```diff
@@ -354,6 +352,12 @@ def test_tested_vs_testable_methods_single_phase_flow(
         ("well_fluid_flux", 0, 2),  # use dim_restriction=2 to ignore well flux
         ("well_flux_equation", 0, 2),  # use dim_restriction=2 to ignore well equation
         ("well_radius", 0.1, None),
+        # Testing of methods with deeper namespace
```
The testing of the fluid mass balance was modified and adapted to the new interface for fluid properties.
The test was extended to cover the standard model implemented (1 phase 1 component).

keileg replied:
This can be resolved from my side

vlipovac reviewed (commented):
Here is a self-review with the essential changes.

All other changes are essentially due to the change in the interface to fluid properties:

`model.fluid_density()` -> `model.fluid.density` and similar.

Also, the access to FluidConstants and SolidConstants is not done by calling anymore, thanks to the data class framework.

keileg commented:
Thanks @vlipovac. I'll get started on this later today.

vlipovac commented on src/porepy/utils/porepy_types.py:
```diff
@@ -35,3 +37,23 @@
     "pp.fracs.fracture_network_2d.FractureNetwork2d",
     "pp.fracs.fracture_network_3d.FractureNetwork3d",
 ]
+
```
I introduced these two types for sake of dynamic creation of compositional variables and phase properties.
Inside `fluid.density()`, the method `phase.density()` is called. While `fluid.density` is a static interface for all other model interfaces, the `phase.density` is created depending on phase context. The static checks and intellisense have now something to work with.

keileg reviewed (commented):
Various comments and questions. Some are marked as 'postpone' and can definitely be ignored while discussing the concepts, but that likely also applies to more methods. There may also be more such minor comments at a later stage (I tried to read with an eye on the big picture in this iteration).

My main impression is that the overall design is sound, I don't see any red flags. There are some questions regarding smaller design choices that should be addressed, or explained, but these should not be show stoppers.

I suggest you have a look at the comments @vlipovac and that we then talk through the changes as a second check that I have not overlooked something. If not, I think @IvarStefansson can have a look quite soon.

vlipovac commented:
@keileg  thanks for the effort!

From your comments I filtered 3 points which are on the more crucial side:

1. `mob_rho_h`: Yes, I would suggest to do that generalization as a next preparation for merger #1236 (to minimize duplicate code in the CF equations)
2. `convert_units`: I agree that the code is cumbersome, due to the attempt to make this merger a bit less breaking. Given your and Ivar's comments about units last Thursday (have a single `Units` instance per simulation and persistently use it), I suggest moving this functionality out of the material constants and into the class `Units`. Each model has a `model.units` attribute, and the usage will be more clear.
3. `fluid` in `MomentumBalance`: This is true, the whole thing works without a fluid and it does not need it. The reason it has it is due to the base `SolutionStrategy` executing `prepare_simulation() -> set_materials() -> create_fluid()`. The last method is mixed in by `FluidMixin` (BTW, also as of now, `MomentumBalance` has a `fluid: FluidConstants` attribute which it does not use). I haven't found a more elegant way without introducing logic, which might be even more conflicting when mixing momentum balance with flow/transport/energy.

Regarding point 3, the core problem is that the fluid creation (constants or general), is on the lowest level in modelling (every solution strategy inherits from the base solution strategy). Any local solution to *not* create a fluid leads to a pitfall when mixing multi-physics models. I.e., the mechanics model must be mixed in last, such that flow, transport & energy override that local solution.
I am not sure if there is an easy, understandable solution which can be programmed, or if it is something users have to be educated about.
Hence my (lazy) solution to just mix it into `MomentumBalance`.

vlipovac commented on src/porepy/compositional/materials.py:
```diff
@@ -0,0 +1,536 @@
+# By providing default values to the fields, not every constant is required. The user
+# is expected to be aware which physics are used in the model.
+@dataclass(frozen=True, kw_only=True)
+class MaterialConstants:
```
Yes. The original implementation performed a conversion when the constant was called, IF it had a physical dimension. Otherwise it just returned the constant.
Now the post init processing takes care of the conversion, IF a dimension is given in SI_units, and we can define constants simply using the dataclass mechanism.

vlipovac commented on src/porepy/models/solution_strategy.py:
```diff
@@ -339,32 +356,48 @@ def reset_state_from_file(self) -> None:
+        # Store the default fluid constants to be accessible by the compositional.Fluid
+        # TODO considering that there is a order in prepare_simulations, is there a better
+        # solution to transfer the fluid constants to the next step in prepare simulations?
+        self.params["_default_fluid_constants"] = fluid
```
Starting with _ to hint to user to not mess with it.
I put it in params such that it is accessible to the FluidMixin when performing create_fluid, and other methods called subsequently. Either this, or I change multiple signatures, but that change requires more code adaptation for the general CF case.

keileg replied:
I think I understand. This is why we went so heavily for mixins in the first place, then we don't have these issues with access of information. I believe the solution is fair provided you clearly mark the intended usage also on relevant places (I cannot find them at the moment)

keileg replied:
Feel free to resolve when you feel you're don, @vlipovac

vlipovac commented on src/porepy/compositional/base.py:
```diff
@@ -337,24 +340,23 @@ def fractions_from_molalities(self, *molalities: FloatType) -> list[FloatType]:
     Note:
-        This class is called 'abstract EoS'. Users can implement any correlations but
-        are encouraged to focus on thermodynamic consistency.
+        This class is called 'abstract EoS', but is in fact not strictly abstract and can be
+        instantiated.
```
I agree. Just EquationOfState?

keileg replied:
Fine with me.

vlipovac commented on src/porepy/models/fluid_mass_balance.py:
```diff
@@ -473,7 +472,9 @@ def bc_values_pressure(self, boundary_grid: pp.BoundaryGrid) -> np.ndarray:
 
         """
-        return self.fluid.pressure() * np.ones(boundary_grid.num_cells)
+        return self.fluid.reference_component.pressure * np.ones(
```
Yes, I also have some troubles accepting pressure or temperature as a material propert. While density (or to be more precise - reference_density) is somehow acceptable, this one would fit better to a phase. But that would require some similar structure X for the phase as well.
FluidConstants -> Component
X -> Phase

keileg replied:
Can be resolved after an explanatory comment has been added

vlipovac commented on tests/models/test_fluid_mass_balance.py:
```diff
@@ -374,8 +378,15 @@ def test_ad_operator_methods_single_phase_flow(
+    method_namespace = method_name.split('.')
+    owner = model_setup
+    # loop top to bottom through namespace to get to the actual method defined on some grids
+    # and returning an operator
```
Yes, the principle of the test stays the same, just dealing with the intermediate layer given by the Fluid class.

keileg commented:
@vlipovac 

> 1. ... Yes, I would suggest to do that generalization as a next preparation ...

Fine with me.

> 2.  ... I suggest moving this functionality out of the material constants and into the class Units.

This makes sense. Do you agree @IvarStefansson?

3. I see the dilemma, and agree that on balance it is better to have an inactive fluid attribute in the momentum balance than to introduce various workarounds. So let's keep the formulation as it is.

vlipovac commented on src/porepy/compositional/materials.py:
```diff
@@ -0,0 +1,536 @@
+
+    This class is intended for 1 species only. Different domains or fluids in the
+    mD-setting require each their own material constants instance in the case of
+    heterogenity.
```
Split and moved documentation to FluidConstants and Solidconstants respectively.

vlipovac commented on src/porepy/compositional/materials.py:
```diff
@@ -0,0 +1,536 @@
+    """
+
+    # NOTE Annotating it as a ClassVar leads to the dataclasss decorator ignoring this in its
+    # machinery. The annotation must not be forgotten in derived classes.
```
Should be resolved throughout the changed files.
(I don't know though why flake8 is not complaining)

vlipovac commented on src/porepy/compositional/materials.py:
```diff
@@ -0,0 +1,536 @@
+
+    thermal_conductivity: number = 1
+
+    normal_thermal_conductivity: number = 1
```
Ordered manually in alphabetical order.

vlipovac commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1222,106 +1250,69 @@ def dependencies_of_phase_properties(
-        else:
+        Important:
+            This method must be overwritten in every flow problem which does not rely
+            on heuristics. The base method returns an empty list for reasons of compatibility
```
Improved documentation to refer to heuristic laws for phase properties.

vlipovac commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1222,106 +1250,69 @@ def dependencies_of_phase_properties(
+                        if self._has_unified_equilibrium:
 
-        """
+                            dependencies +=  [self.temperature] + [
```
Added documentation of the code snippet and intra-sphinx links to the implementation of the boolean properties.

vlipovac commented on src/porepy/compositional/base.py:
```diff
@@ -337,24 +340,23 @@ def fractions_from_molalities(self, *molalities: FloatType) -> list[FloatType]:
-        able to pick up the dependencies and call :meth:`compute_phase_state` with
-        the right values.
+        This is by intention such that phases can be created with the abstract EoS, in a
+        simulation setting which uses heuristics. The method :meth:`compute_phase_properties`
```
Clarified docs.

vlipovac commented on src/porepy/models/protocol.py:
```diff
@@ -0,0 +1,816 @@
+    class PorePyModel(Protocol):
+        """This is an empty placeholder of the protocol, used mainly for type hints."""
+
+        def __init__(self, *args, **kwargs) -> None:
```
@keileg @Yuriyzabegaev Marking this change here, since it fixes something we do not know of as of now.

vlipovac commented:
Added changes as per #1247 

I rebased the branch onto the recent version of develop which includes Protocols.
A `models.protocol.FluidProtocol` is included now.
I think we can have a second round of review @keileg @IvarStefansson .
Also, most of the remarks made by Eirik should be fixed now.

vlipovac commented on src/porepy/models/protocol.py:
```diff
@@ -0,0 +1,825 @@
+
+Note that the protocol framework is accessed by static type checkes only!
+
+Warning:
```
Adding this warning to reflect the insights gained with @Yuriyzabegaev 
The runtime `PorePyModel` class does not inherit from `Protocol` anymore.

keileg commented on src/porepy/viz/data_saving_model_mixin.py:
```diff
@@ -10,52 +10,24 @@
     def save_data_time_step(self) -> None:
-        """Export the model state at a given time step, and log time.
-        The options for exporting times are:
-            * None: All time steps are exported
```
I don't think we want to delete the docstrings from these methods, even though they are also present in the protocol. Do you agree @Yuriyzabegaev?

keileg replied:
This applies to several methods in this file.

Yuriyzabegaev replied:
Yes, the current convention is to duplicate docstrings. This is something we can mention on the next user meeting.

vlipovac replied:
This change was not intended, may have been caused by the messy rebase. I will check for other missing docstrings.

keileg commented on src/porepy/models/fluid_property_library.py:
```diff
@@ -0,0 +1,338 @@
+    mobility_keyword: str
+    """Keyword for the discretization of the mobility. Normally provided by a mixin of
+    instance :class:`~porepy.models.SolutionStrategy`.
+
```
Let's try not to introduce new todos unless they are strictly necessary.

keileg reviewed (commented):
Some minor comments before I had to run for the weekend. I think most of my concerns are taken care of, I'll try to resolve comments when I have the time.

Navigating the changes in difficult due to the protocol changes being included. I suggest you consider to formally close this PR (after we have an overview of which comments have been taken care of), and then open a new one before Ivar has a look - that should make things easier for him. does this make sense?

keileg commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1222,106 +1228,102 @@ def dependencies_of_phase_properties(
+            This method must be overwritten in every flow problem which relies on e.g.,
+            :class:`~porepy.numerics.ad.surrogate_operator.SurrogateFactory` for
+            externally computed values. The framework requires information about the
+            input arguments of the externally computes property values (primary
```
```suggestion
            input arguments of the externally computed property values (primary
```

keileg commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1222,106 +1228,102 @@ def dependencies_of_phase_properties(
+            Let's assume a general mixture with multiple components and phases, which
+            uses the :class:`CompositionalVariables`, and either
+            :attr:`_MixtureDOFHandler._has_unified_equilibrium` or
+            :attr:`_MixtureDOFHandler._has_equilibrium` defined. I.e., the
```
```suggestion
            :attr:`_MixtureDOFHandler._has_equilibrium` defined. i.e., the
```

keileg commented on src/porepy/models/energy_balance.py:
```diff
@@ -673,14 +675,13 @@ def reference_temperature(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
         """
-        t_ref = self.fluid.temperature()
-        assert t_ref == self.solid.temperature()
+        t_ref = self.fluid.reference_component.temperature
```
Can be resolved after a short explanation has been added

vlipovac commented:
Closing this PR and re-submitting in #1252 

Reasons include a confusing git history due to a messy rebase.
